### PR TITLE
CORDA-3022 Add wildcard RPC permissions

### PR DESCRIPTION
**Pull request by StefanIliev545**
Added maintainer permission group and a test hack where "." are replaced with ":" in permission strings to support * (which may not be feasible running with a db backend for rpc authentication)

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/internal/security/RPCSecurityManagerImpl.kt [hunk 1]
```diff
@@ -143,12 +142,26 @@ private object RPCPermissionResolver : PermissionResolver {
...
+            "isFlowsDrainingModeEnabled",
+            "nodeInfo",
+            "currentNodeTime",
+            "killFlow",
+            "isWaitingForShutdown",
+            "shutdown",
+            "terminate",
+            "gracefulShutdown",
+            "clearNetworkMapCache"
+    )
```
Comment (szymonsztuka) [inline node/src/main/kotlin/net/corda/node/internal/security/RPCSecurityManagerImpl.kt]: This is more a disscussion,  maybe each permission should be assigned to at least one predefined role  ("maintainer" , "invokerpc",  "node-user"?)

More permsissons for "maintenier":
-  all network related commands like `networkMapSnapshot`
 - readonly for attachments  (`attachmentExists`), more ermissions to upload attachments?
- `waitUntilNetworkReady`

File: node/src/main/kotlin/net/corda/node/internal/security/RPCSecurityManagerImpl.kt [hunk 2]
```diff
@@ -114,15 +114,14 @@ private class RPCPermission : DomainPermission {
      * @param methods Set of allowed RPC methods
      * @param target  An optional "target" type on which methods act
      */
-    constructor(methods: Set<String>, target: String? = null) : super(methods, target?.let { setOf(it) })
+    constructor(methods: Set<String>, target: String? = null) : super(methods, target?.let { setOf(it.replace(".", ":")) })
```
Comment (szymonsztuka) [inline node/src/main/kotlin/net/corda/node/internal/security/RPCSecurityManagerImpl.kt]: This seems the only sensible way to allow wildcard, not sure if we should amend our docs in  https://docs.corda.net/tutorial-clientrpc-api.html#security and change "." to ":"  to explicit say it allows wildcards (and for backwards compatibility "." are replaced with ":" anyway)

File: node/src/main/kotlin/net/corda/node/internal/security/RPCSecurityManagerImpl.kt [hunk 3]
```diff
@@ -175,6 +188,9 @@ private object RPCPermissionResolver : PermissionResolver {
                 // Leaving empty set of targets and actions to match everything
                 return RPCPermission()
             }
+            ACTION_MAINTAINER-> {
```
Comment (szymonsztuka) [inline node/src/main/kotlin/net/corda/node/internal/security/RPCSecurityManagerImpl.kt]: the docs should mention new "maintainer" role in "Note" part of https://docs.corda.net/tutorial-clientrpc-api.html#security (not ideal, but to have a trace of it for now)

Comment (anthonykeenan) [reply node/src/main/kotlin/net/corda/node/internal/security/RPCSecurityManagerImpl.kt]: This isn't the 'full' PR just a bit of a demo to see what we can do with what's existing - obviously the docs would need a rewrite.

However I am intested in how the RPC permissions would be setup in the `node.conf` or DB, and whether we can combine them. Say if we installed a 'maintainer' cordapp for checkign flows, could we create a user with MAINTAINER + invoke for that specific flow? For the meeting next tuesday I'd like to see an example node set up to use this, using `node.conf` to specify permissions, and also one with perms in the db

**Comment by anthonykeenan** [comment]
I've updated this now to only do the wildcard work and added some documentation in - if you could review @szymonsztuka that would be great

File: node/src/main/kotlin/net/corda/node/internal/security/RPCSecurityManagerImpl.kt [hunk 4]
```diff
@@ -114,15 +114,14 @@ private class RPCPermission : DomainPermission {
...
-    constructor(methods: Set<String>, target: String? = null) : super(methods, target?.let { setOf(it) })
+    constructor(methods: Set<String>, target: String? = null) : super(methods, target?.let { setOf(it.replace(".", ":")) })
 
 
     /**
      * Default constructor instantiate an "ALL" permission
      */
     constructor() : super()
 }
-
```
Comment (szymonsztuka) [inline node/src/main/kotlin/net/corda/node/internal/security/RPCSecurityManagerImpl.kt]: Unnecessary new line deletion.

**Pull request merged**